analyzer/src/steps/typing.rs

```diff
@@ -260,29 +358,33 @@ fn ascribe_assign(
     diagnostics: &mut Vec<Diagnostic>,
     state: TypingState,
 ) -> TypedExpr {
-    let rhs = match assign.operator {
-        AssignOperator::Assign => ascribe_types(
-            exploration,
-            links,
-            diagnostics,
-            &assign.value,
-            state.with_local_type(),
-        ),
-        operator => {
-            let binary = Expr::Binary(BinaryOperation {
-                left: assign.left.clone(),
-                op: BinaryOperator::try_from(operator).expect("Invalid assign operator"),
-                right: assign.value.clone(),
-            });
-            ascribe_types(
-                exploration,
-                links,
-                diagnostics,
-                &binary,
-                state.with_local_type(),
-            )
-        }
-    };
+    macro_rules! ascribe_rhs {
+        ($expected_type:expr) => {
+            match assign.operator {
+                AssignOperator::Assign => ascribe_types(
+                    exploration,
+                    links,
+                    diagnostics,
+                    &assign.value,
+                    state.with_local_type().with_expected_type($expected_type),
+                ),
+                operator => {
+                    let binary = Expr::Binary(BinaryOperation {
+                        left: assign.left.clone(),
+                        op: BinaryOperator::try_from(operator).expect("Invalid assign operator"),
+                        right: assign.value.clone(),
+                    });
+                    ascribe_types(
+                        exploration,
+                        links,
+                        diagnostics,
+                        &binary,
+                        state.with_local_type(),
+                    )
+                }
+            }
+        };
+    }
```

Comment:
```suggestion
    let actual_type_ref = links
        .env()
        .get_raw_symbol(assign.left.segment())
        .map(|symbol| {
            exploration
                .ctx
                .get(links.relations, links.source, symbol)
                .unwrap()
                .type_ref
        });
    let rhs = match assign.operator {
        AssignOperator::Assign => ascribe_types(
            exploration,
            links,
            diagnostics,
            &assign.value,
            state.with_local_type().with_expected_type(actual_type_ref),
        ),
        operator => {
            let binary = Expr::Binary(BinaryOperation {
                left: assign.left.clone(),
                op: BinaryOperator::try_from(operator).expect("Invalid assign operator"),
                right: assign.value.clone(),
            });
            ascribe_types(
                exploration,
                links,
                diagnostics,
                &binary,
                state.with_local_type(),
            )
        }
    };
```
Avoid big macros that lose syntax highlighting and that can be avoided here.

Reply:
it causes code duplication

Reply:
Macros are just expanded. The point here is that there is no need for two calls to a `ascribe_rhs` macro.